background.js

```diff
@@ -10,21 +21,152 @@ chrome.runtime.onMessage.addListener(async (message) => {
+  const { enable_auto_archive } = await chrome.storage.local.get(['enable_auto_archive']);
+  console.debug(`[Auto-Archive Debug] enable_auto_archive setting: ${enable_auto_archive}`);
+
+  if (tabUpdateListener) {
+    try {
+      console.debug('[Auto-Archive Debug] Removing existing tab update listener');
+      chrome.tabs.onUpdated.removeListener(tabUpdateListener);
+      tabUpdateListener = null;
+    } catch (error) {
+      console.error('Error removing listener:', error);
+    }
+  }
+
+  if (enable_auto_archive) {
+    console.debug('[Auto-Archive Debug] Auto-archive is enabled, checking permissions');
+    const hasPermission = await chrome.permissions.contains({ permissions: ['tabs'] });
+    console.debug(`[Auto-Archive Debug] Has tabs permission: ${hasPermission}`);
+
+    if (hasPermission) {
+      tabUpdateListener = async (tabId, changeInfo, tab) => {
+        console.debug(`[Auto-Archive Debug] Tab updated - tabId: ${tabId}, status: ${changeInfo.status}, url: ${tab?.url}`);
+
+        // Only process when the page has completed loading
+        if (changeInfo.status === 'complete' && tab.url) {
+          console.debug(`[Auto-Archive Debug] Tab load complete, checking if URL should be archived: ${tab.url}`);
+
+          const shouldArchive = await shouldAutoArchive(tab.url);
+          console.debug(`[Auto-Archive Debug] shouldAutoArchive result: ${shouldArchive}`);
+
+          if (shouldArchive) {
+            console.log('Auto-archiving URL:', tab.url);
+
+            const entry = new Entry(
+              tab.url,
+              ['auto-archived'],
+              tab.title,
+              tab.favIconUrl,
+            );
+
+            console.debug('[Auto-Archive Debug] Created new entry, saving to storage');
+            const { entries = [] } = await chrome.storage.local.get('entries');
+            entries.push(entry);
+            await chrome.storage.local.set({ entries });
+            console.debug('[Auto-Archive Debug] Entry saved to local storage');
+
+            try {
+              console.debug(`[Auto-Archive Debug] Calling addToArchiveBox with URL: ${entry.url}, tags: ${entry.tags.join(',')}`);
+              await addToArchiveBox([entry.url], entry.tags.join(','));
+              console.log(`Automatically archived ${entry.url}`);
+            } catch (error) {
+              console.error(`Failed to automatically archive ${entry.url}: ${error.message}`);
+            }
+          }
+        }
+      };
+
+      console.debug('[Auto-Archive Debug] Adding tab update listener');
+      chrome.tabs.onUpdated.addListener(tabUpdateListener);
+      console.log('Auto-archiving enabled with tabs permission');
+    } else {
+      console.log('Tabs permission not granted, auto-archiving disabled');
+      console.debug('[Auto-Archive Debug] No tabs permission, disabling auto-archive setting');
+      // Reset the toggle if permission was not granted
+      chrome.storage.local.set({ enable_auto_archive: false });
```

Comment:
can you hoist this to the top of the function to so we dont have to nest all the logic above

```javascript
if (!hasPermission) {
    ...
}

...
```